boxsdk/util/translator.py

```diff
@@ -116,17 +134,45 @@ def get(self, key, default=None):
             default = BaseObject
         return super(Translator, self).get(key, default)
 
-    def translate(self, type_name):
+    def translate(self, session, response_object):
         """
-        Get the box object class associated with the given type name.
-
-        :param type_name:
-            The type name to be translated.
-        :type type_name:
-            `unicode`
-        :rtype:   :class:`BaseAPIJSONObjectMeta`
+        Translate a given API response object into SDK classes, rescursively translating any subobjects.
+
+        :param session:
+            The SDK session to use for any objects that require a session (i.e. classes that make API calls)
+        :type session:
+            class:`Session`
+        :param response_object:
+            The JSON response object from the API, which will be translated
+        :type response_object:
+            `dict`
+        :return:
+            The translated object
         """
-        return self.get(type_name)
+
+        if isinstance(response_object, dict):
+            for key in response_object:
+                if isinstance(response_object[key], dict):
+                    response_object[key] = self.translate(session, response_object[key])
+                elif isinstance(response_object[key], list):
+                    response_object[key] = [self.translate(session, o) for o in response_object[key]]
+
+            # Try to translate any API object with a `type` property, except for metadata instances
+            # The $type value in metadata instances isn't directly usable, so we avoid it altogether
+            # NOTE: Currently, we represent metadata as just a `dict`, so there's no need to translate it anyway
+            if 'type' in response_object and '$type' not in response_object:
+                object_class = self.get(response_object.get('type', ''))
+                param_values = {
+                    'session': session,
+                    'response_object': response_object,
+                    'object_id': _get_object_id(response_object),
+                }
+                # NOTE: getargspec() is deprecated, and should be replaced by inspect.signature() when 2.7 support drops
```

Comment:
I hope 2.7 support doesn't outlast support for `getargspec`, but it might. It could be safer or more correct to use `inspect.signature`, falling back to `getargspec` when unavailable.

I think the `mock` library does something similar.